### ISSUE TITLE
Slow on V100

### ISSUE THREAD

**Issue by Louis-y-nlp** [mosaicml/llm-foundry#153]
Hi teams, I'm fine-tuning with 6 V100 GPUs. The fine-tuning process is extremely slow for me. I'm using fp16 and attn_impl: torch, with a global_train_batch_size of 12 and device_train_microbatch_size automatically set to 2 (device_train_microbatch_size: auto). Even after 15 hours, I haven't finished training one-third of an epoch (500k rows of data). Did I miss anything?

**Comment by alextrott16** [comment]
Can you confirm that you are running `composer train.py <yaml>` and not `python train.py <yaml>`?

And if you could share as much of the run YAML as you are willing to, that would be appreciated!

**Comment by Louis-y-nlp** [comment]
here is my test.yaml and im running `CUDA_VISIBLE_DEVICES=0,1,2,3,4,5 composer train.py test.yaml`
max_seq_len: 2048
global_seed: 17

# Run Name
run_name: test # If left blank, will be read from env var $COMPOSER_RUN_NAME

model:
  name: hf_causal_lm
  pretrained: true
  pretrained_model_name_or_path: /data/yly/models/generate_models/mosaicml_mpt-7b-instruct
  config_overrides:
    attn_config:
      attn_impl: torch
      # Set this to `true` if using `train_loader.dataset.packing_ratio` below
      attn_uses_sequence_id: false

# Tokenizer
tokenizer:
  name: /data/yly/models/generate_models/mosaicml_mpt-7b-instruct
  kwargs:
    model_max_length: ${max_seq_len}

dataset: &hf_dataset
  hf_name: json
  hf_kwargs:
    data_dir: /data/yly/data/mpt_data/

# Dataloaders
train_loader:
  name: finetuning
  dataset:
    <<: *hf_dataset
    split: train
    tokenizer_name: ${tokenizer_name}
    max_seq_len: ${max_seq_len}
    allow_pad_trimming: false
    decoder_only_format: true
    # # Use `python llmfoundry/data/packing.py --yaml-path /path/to/this/yaml/ ...`
    # # to profile this run's optimal packing_ratio as it depends on GPU count,
    # # batch size, sequence length
    # packing_ratio:
    shuffle: true
  drop_last: true
  num_workers: 6
  pin_memory: false
  prefetch_factor: 2
  persistent_workers: true
  timeout: 0

eval_loader:
  name: finetuning
  dataset:
    <<: *hf_dataset
    split: validation
    tokenizer_name: ${tokenizer_name}
    max_seq_len: ${max_seq_len}
    allow_pad_trimming: false
    decoder_only_format: true
    # packing_ratio:
    shuffle: true
  drop_last: true
  num_workers: 8
  pin_memory: false
  prefetch_factor: 2
  persistent_workers: true
  timeout: 0

# Optimization
scheduler:
  name: linear_decay_with_warmup  # linear no warmup is HF default which dolly used
  t_warmup: 50ba  # add some warmup though, seems to help with MPT
  alpha_f: 0

optimizer:
  # Based on Dolly
  name: decoupled_adamw
  lr: 5.0e-6
  betas:
  - 0.9
  - 0.999
  eps: 1.0e-8
  weight_decay: 0

algorithms:
  gradient_clipping:
    clipping_type: norm
    clipping_threshold: 1.0

max_duration: 2ep  # 2-3 epochs seems like the sweet spot
eval_interval: 1ep
# eval_subset_num_batches: -1
eval_first: true
global_train_batch_size: 12  # somewhere in the 6-8 * numgpus range seems good

# System
seed: ${global_seed}
device_eval_batch_size: 1
# device_train_microbatch_size: 8
device_train_microbatch_size: auto
#precision: amp_bf16
precision: amp_fp16

# FSDP
fsdp_config:
  sharding_strategy: FULL_SHARD
  mixed_precision: PURE
  activation_checkpointing: true
  activation_checkpointing_reentrant: false
  #activation_cpu_offload: false
  activation_cpu_offload: true
  limit_all_gathers: true
  verbose: false

# Logging
progress_bar: false
log_to_console: true
console_log_interval: 1ba

callbacks:
  speed_monitor:
    window_size: 10
  lr_monitor: {}
  memory_monitor: {}
  runtime_estimator: {}

# loggers:
#   wandb: {}

# Checkpoint to local filesystem or remote object store
save_interval: 500ba
save_num_checkpoints_to_keep: 4  # Important, this cleans up checkpoints saved to DISK
save_folder: ./{run_name}/checkpoints
# save_folder: s3://my-bucket/my-folder/{run_name}/checkpoints

**Comment by Louis-y-nlp** [comment]
I understand that the format is not ideal, but I don't know how to upload the correct format. I apologize for the inconvenience.

**Comment by abhi-mosaic** [comment]
Hi @Louis-y-nlp , I think this is a duplicate of #114 and I left [a couple of comments there](https://github.com/mosaicml/llm-foundry/issues/114#issuecomment-1569335882)
